lib/containers/utils.pm

```diff
@@ -315,4 +315,85 @@ sub reset_container_network_if_needed {
     }
 }
 
+sub test_volumes {
+    my %args = @_;
+    my $runtime = $args{runtime};
+
+    # From https://docs.docker.com/storage/bind-mounts/
+    # The --mount flag does not support z or Z options for modifying selinux labels.
+    my $Z = $runtime eq "podman" ? ",Z" : "";
+
+    # Create a directory with the sticky-bit set because of rootless podman
+    assert_script_run('mkdir -m 1777 test');
+
+    # Create Dockerfile with VOLUME defined
+    assert_script_run(q{echo -e 'FROM busybox\nVOLUME /test' > test/Dockerfile});
+
+    # Build image and load it
+    assert_script_run("$runtime build -t test -f test/Dockerfile test/");
+
+    assert_script_run('touch test/test');
+
+    # Test --volumes-from
+    assert_script_run("$runtime run -d --name test test");
+    assert_script_run("$runtime run --rm --volumes-from test test ls /test | grep test");
+    assert_script_run("$runtime rm -vf test");
+
+    assert_script_run("$runtime run -d --name test -v \$PWD/test:/test:Z test");
+    # NOTE: When https://github.com/containers/podman/issues/19529 is solved we must add a version check here
+    my $ret = script_run("$runtime run --rm --volumes-from test test ls /test/test");
+    record_soft_failure("gh#19529 - Unexpected error with --volumes-from") if ($ret != 0 && $runtime == "podman");
+    assert_script_run("$runtime rm -vf test");
+
+    # Test --volume option with directory (read-only)
+    assert_script_run("! $runtime run --rm --volume \$PWD/test:/test:ro,Z test rm /test/test");
+    assert_script_run("test -f test/test");
+
+    # Equivalent --mount option to above
+    assert_script_run("! $runtime run --rm --mount type=bind,source=\$PWD/test,destination=/test,readonly$Z test rm /test/test");
+    assert_script_run("test -f test/test");
+
+    assert_script_run("rm test/test");
+
+    # Test --volume option with directory (read-write)
+    assert_script_run("$runtime run --rm --volume \$PWD/test:/test:Z test touch /test/test");
```

Comment:
It might be a good idea to put the mount flags into some data structure

Reply:
> It might be a good idea to put the mount flags into some data structure

What do you suggest while still maintaining the code readable?

Reply:
I do not have anything concrete at the moment. The engine flags for podman and docker are not 1:1, and currently it is slightly complicated to follow. As somewhere are hardwired and some are coming from a variable.

Reply:
> I do not have anything concrete at the moment. The engine flags for podman and docker are not 1:1, and currently it is slightly complicated to follow. As somewhere are hardwired and some are coming from a variable.

Let's do that when we add testing for more esoteric options.  But first we'd have to be sure which ones and check when they were introduced.  For now I think we're covering the proverbial 80% of cases.

Reply:
I agree with Ricardo here because there is no real good way of refactoring it as of now. Or do you have a good idea?